File: lesson-4/ruby/max_counters.rb
def solution(n, a)
  counters = Array.new(n, 0)
  increment_previous = 0
  increment_max = 0
  a.each do |i|
    pos = i - 1;
    if i > n
      increment_max = increment_previous
    else
      if increment_max > counters[pos]
        counters[pos] = increment_max + 1
      else
        counters[pos] += 1
      end
      if counters[pos] > increment_previous
        increment_previous = counters[pos]
      end
    end
  end
  counters.map { |x| x < increment_max ? increment_max : x }
end

File: lesson-4/js/maxCounters.js
module.exports = function solution(n, a) {

  let counters = new Array(n).fill(0);
  let incrementMax = 0;
  let incrementLast = 0;

  a.forEach(num => {
    let index = num - 1;
    if (num > n)
      incrementMax = incrementLast;
    else {
      if (incrementMax > counters[index])
        counters[index] = incrementMax + 1;
      else
        counters[index] += 1;
      if (counters[index] > incrementLast)
        incrementLast = counters[index];
    }
  });

  return counters.map(num => num < incrementMax ? incrementMax : num);

};